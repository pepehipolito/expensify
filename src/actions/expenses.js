import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt};

    // We need to return the promise so we can test it. This requires us to have a connection to the
    // DB for testing, though. If that is not available the tests will fail.
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  }
}

// REMOVE_EXPENSE
export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, payload) => ({
  type: 'EDIT_EXPENSE',
  id,
  payload
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// export const startSetExpenses;
export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then(snapshot => {
      const expenses = [];

      snapshot.forEach(childSnapshot => {
        expenses.push({
          id: parseInt(childSnapshot.key),
          ...childSnapshot.val()
        })
      });

      dispatch(setExpenses(expenses));
    });
  }
}
