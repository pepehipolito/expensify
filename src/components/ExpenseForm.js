import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    const { expense } = this.props;

    this.state = {
      description:      expense ? expense.description : '',
      note:             expense ? expense.note : '',
      amount:           expense ? (expense.amount / 100).toString() : '',
      createdAt:        expense ? moment(expense.createdAt) : moment(),
      calendarFocused:  false,
      error:            ''
    };
  }

  onDescriptionChange = (e) => {
    // We could have done something like this, but it would have given us errors because the target
    // doesn't exist at definition time:
    //  this.setState(() => ({description: e.target.value}))
    // In order to avoid the error and make it work we would need to persist the event first:
    //  e.persist();
    //  this.setState(() => ({description: e.target.value}))
    // But it's just easier this way:
    const description = e.target.value;
    this.setState(() => ({description}))
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/) ) {
      this.setState(() => ({amount}));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}));
    }
  };

  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}))
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (this.state.description && this.state.amount) {
      this.setState(() => ({error: ''}));
      // amount: we need to parse it as a float because it is stored in the state in string format
      // and we need to mulitply it by 100 to convert it to cents.
      // createdAt: we need to convert it to a millisecond based timestamp with moment's 'valueOf'
      // method because it is a 'moment' instance.
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    } else {
      this.setState(() => ({error: 'Please provide description and amount.'}));
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <h3>{this.state.error}</h3>}

        <form onSubmit={this.onSubmit}>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />

          <input
            type='text'
            placeholder='Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={(day) => false}
          />

          <textarea
            placeholder='Add a note for your expense (optional)'
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>{this.props.action} Expense</button>
        </form>
        {this.props.action === 'Update' && <button onClick={this.props.onRemove}>Remove</button>}
      </div>
    );
  }
}
