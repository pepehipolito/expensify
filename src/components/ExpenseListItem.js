import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

//FIXME: The id value is always NaN, so data does not show and record cannot be removed.
const ExpenseListItem = ({id, description, amount, createdAt, dispatch}) => (
  <div>
    <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
      <p>
        {numeral(amount / 100).format('$,0.00')}
        -
        {moment(createdAt).format('MMMM Do, YYYY')}
      </p>
  </div>
);

export default ExpenseListItem;
