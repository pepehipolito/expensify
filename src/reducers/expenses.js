const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state, action.expense
      ]
    case 'REMOVE_EXPENSE':
      // Here we are destructuring the expense element to obtain the id value in the argument to 'filter'.
      return state.filter(({id}) => (id !== action.id ));
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {...expense, ...action.payload};
        } else {
          return expense;
        }
      });
    case 'SET_EXPENSES':
      return action.expenses;
    default:  // This gets called when setting up the store, so we need a default 'state' parameter value.
      return state;
  }
};
