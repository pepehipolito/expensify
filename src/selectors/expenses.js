import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    // If no startDate/endDate filtering set we accept all dates.
    const startDateMatch  = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day')  : true;
    const endDateMatch    = endDate   ? endDate.isSameOrAfter(createdAtMoment, 'day')     : true;
    // If no text filtering set (text.length === 0) we accept all texts.
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    switch (sortBy) {
      case 'date':
        return a.createdAt < b.createdAt ? 1 : -1;
      case 'amount':
        return a.amount < b.amount ? 1 : -1;
      default:
        return a;
    }
  });
};
