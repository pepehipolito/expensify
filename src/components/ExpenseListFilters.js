import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocus: null
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocus) => {
    this.setState(() => ({calendarFocus}));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortByChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
        <input
          type='text'
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />

        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortByChange}
        >
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocus}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({filters: state.filters});
const mapDispatchToProps = (dispatch, props) => ({
  setStartDate:   (date)  => dispatch(setStartDate(date)),
  setEndDate:     (date)  => dispatch(setEndDate(date)),
  setTextFilter:  (text)  => dispatch(setTextFilter(text)),
  sortByDate:     ()      => dispatch(sortByDate()),
  sortByAmount:   ()      => dispatch(sortByAmount())
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
