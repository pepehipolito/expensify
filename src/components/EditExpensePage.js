import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Edit Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            action='Update'
            onSubmit={this.onSubmit}
            onRemove={this.onRemove}
            expense={this.props.expense}
          />
        <button className='button button__secondary' onClick={this.onRemove}>Remove Expense</button>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense:   (expenseId, expense) => dispatch(startEditExpense(expenseId, expense)),
  startRemoveExpense: expenseId => dispatch(startRemoveExpense(expenseId))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
