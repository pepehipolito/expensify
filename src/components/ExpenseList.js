import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';

// We could have named this whatever we wanted. The file has a unique unnamed exported function.
import selectExpenses from '../selectors/expenses';

// We export this one to be able to test it. We test the unconnected version.
export const ExpenseList = (props) => (
  <div className='content-container'>
    <div className='list-header'>
      <div className='show-for-mobile'>Expenses</div>
      <div className='show-for-desktop'>Expense</div>
      <div className='show-for-desktop'>Amount</div>
    </div>
    <div className='list-body'>
      {
        props.expenses.length === 0 ? (
          <div className='list-item list-item__message'>
            <span>No expenses</span>
          </div>
        ) : (
          props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
        )
      }
    </div>
  </div>
);

// This will map the state values and return an object that will be used as props for the regular
// component (ExpenseList). The function is implicitely returning the object.
const mapStateToProps = (state) => ({expenses: selectExpenses(state.expenses, state.filters)});

// connect() accepts a function (mapStateToProps here) that returns an object that will be used as
// props to the regular component we are connecting to (ExpenseList here).
// When a component is connected to the redux store it becomes "reactive", it gets re-rendered when
// the store changes. connect() makes the component a subscriber/listener to changes in the store.
// Since we are returning above the entire contents of state we could have rewritten the following
// line as: export default connect((state) => state)(ExpenseList);
export default connect(mapStateToProps)(ExpenseList);
