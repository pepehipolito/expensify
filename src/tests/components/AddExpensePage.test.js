import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let addExpenseSpy, historySpy, wrapper;

beforeEach(() => {
  addExpenseSpy = jest.fn();
  historySpy    = { push: jest.fn() };
  wrapper       = shallow(<AddExpensePage startAddExpense={addExpenseSpy} history={historySpy} />);
});

test('renders AddExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test("sets the form action to 'Add'", () => {
  const action = wrapper.find('ExpenseForm').prop('action');
  expect(action).toBe('Add');
});

test('handles onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(historySpy.push).toHaveBeenLastCalledWith('/');
  expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[1]);
});
