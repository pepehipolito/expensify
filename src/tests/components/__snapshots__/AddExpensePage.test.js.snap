// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AddExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      action="Add"
      onSubmit={[Function]}
    />
  </div>
</div>
`;
